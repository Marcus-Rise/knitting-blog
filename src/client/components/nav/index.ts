export { Nav } from "./nav.component";
export type { INavLink } from "./nav-link.interface";
