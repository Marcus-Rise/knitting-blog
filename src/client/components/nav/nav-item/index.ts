export { NavItem } from "./nav-item.component";
