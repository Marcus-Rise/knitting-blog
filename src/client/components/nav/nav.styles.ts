import styled, { css } from "styled-components";
import { media } from "../../styles";

const Root = styled.nav`
  border-top: 1px dashed #000000;
  border-bottom: 1px dashed #000000;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;

  ${media.md} {
    border: none;
    padding: 0;
  }
`;

const Ul = styled.ul`
  list-style: none;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0;

  ${media.md} {
    flex-direction: column;
    margin: 0;
  }
`;

const Li = styled.li`
  ${media.md} {
    padding-bottom: 1rem;
    width: 60%;
    text-align: center;

    &:not(:last-child) {
      border-bottom: 1px dashed black;
      margin-bottom: 1rem;
    }
  }
`;

const MobileToggle = styled.div`
  display: none;
  justify-content: flex-end;

  ${media.md} {
    display: flex;
  }
`;

const ItemsContainerShow = css`
  display: flex !important;
`;

const ItemsContainer = styled.div<{ showMobileMenu?: boolean }>`
  ${media.md} {
    display: none;
    flex-direction: column;
    position: fixed;
    top: 0;
    right: 0;
    z-index: 901;
    background-color: white;
    width: 80vw;
  }

  ${(props) => props.showMobileMenu && ItemsContainerShow};
`;

const CloseButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export { Root, Ul, Li, MobileToggle, ItemsContainer, CloseButtonContainer };
