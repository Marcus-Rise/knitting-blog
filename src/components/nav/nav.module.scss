@use "../../styles/typography";
@use "../../styles/grid";


.root {
  $border: 1px dashed #000000;
  border-top: $border;
  border-bottom: $border;

  $padding: 0.5rem;
  padding-top: $padding;
  padding-bottom: $padding;


  @include grid.media-mobile-only {
    border: none;
    $padding: 1rem;
    padding: $padding;
  }
}

.mobileToggle {
  display: flex;
  justify-content: flex-end;

  @include grid.media-mobile-up {
    display: none;
  }
}

.itemsContainer {
  @include grid.media-mobile-only {
    display: none;

    position: fixed;
    top: 0;
    right: 0;
    z-index: 901;

    background-color: white;

    width: 80vw;
  }
}

.ul {
  list-style: none;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0;

  @include grid.media-mobile-only {
    flex-direction: column;
  }
}

.li {
  @include typography.light;

  font-size: 1.25rem;
  text-transform: uppercase;

  a {
    text-decoration: none;
    color: inherit;
  }

  @include grid.media-mobile-only {
    margin-bottom: 1.5rem;
  }
}


