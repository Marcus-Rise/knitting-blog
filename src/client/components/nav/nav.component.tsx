import type { FC } from "react";
import React, { useCallback, useMemo, useState } from "react";
import type { INavLink } from "./nav-link.interface";
import { Overlay } from "../overlay";
import { MobileMenuButton } from "../mobile-menu-button";
import { MobileMenuCloseButton } from "../mobile-menu-close-button";
import { NavItem } from "./nav-item";
import { CloseButtonContainer, ItemsContainer, Li, MobileToggle, Root, Ul } from "./nav.styles";
import { Container } from "../container";

interface IProps {
  items: ReadonlyArray<INavLink>;
}

const Nav: FC<IProps> = (props) => {
  const [showMobileMenu, setShowMenu] = useState<boolean>(false);
  const openMenu = useCallback((): void => setShowMenu(true), []);
  const closeMenu = useCallback((): void => setShowMenu(false), []);

  const items = useMemo(
    () =>
      props.items.map((i) => (
        <Li key={i.title} onClick={closeMenu}>
          <NavItem {...i} />
        </Li>
      )),
    [closeMenu, props.items],
  );

  return (
    <Root>
      <MobileToggle>
        <MobileMenuButton onClick={openMenu} />
      </MobileToggle>
      {showMobileMenu && <Overlay onClose={closeMenu} />}
      <ItemsContainer as={Container} showMobileMenu={showMobileMenu}>
        {showMobileMenu && (
          <CloseButtonContainer>
            <MobileMenuCloseButton onClick={closeMenu} />{" "}
          </CloseButtonContainer>
        )}
        <div>
          <Ul>{items}</Ul>
        </div>
      </ItemsContainer>
    </Root>
  );
};

export { Nav };
