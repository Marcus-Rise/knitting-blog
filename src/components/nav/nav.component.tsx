import React, { useCallback, useMemo, useState } from "react";
import styles from "./nav.module.scss";
import type { INavLink } from "./nav-link.interface";
import { Overlay } from "../overlay";
import { MobileMenuButton } from "../mobile-menu-button";
import { MobileMenuCloseButton } from "../mobile-menu-close-button";
import classnames from "classnames";
import { NavItem } from "./nav-item";

interface IProps {
  items: Array<INavLink>;
}

const Nav: React.FC<IProps> = (props) => {
  const [showMobileMenu, setShowMenu] = useState<boolean>(false);
  const openMenu = useCallback((): void => setShowMenu(true), []);
  const closeMenu = useCallback((): void => setShowMenu(false), []);

  const items = useMemo(
    () =>
      props.items.map((i) => (
        <li className={styles.li} key={i.title} onClick={closeMenu}>
          <NavItem {...i} />
        </li>
      )),
    [closeMenu, props.items],
  );

  return (
    <nav className={styles.root}>
      <div className={styles.mobileToggle}>
        <MobileMenuButton onClick={openMenu} />
      </div>
      {showMobileMenu && <Overlay onClose={closeMenu} />}
      <div className={classnames(styles.itemsContainer, { [styles.itemsContainerShow]: showMobileMenu })}>
        {showMobileMenu && (
          <div className={styles.closeButtonContainer}>
            <MobileMenuCloseButton onClick={closeMenu} />{" "}
          </div>
        )}
        <div className="container">
          <ul className={styles.ul}>{items}</ul>
        </div>
      </div>
    </nav>
  );
};

export { Nav };
