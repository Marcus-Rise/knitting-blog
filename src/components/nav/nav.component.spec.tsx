import { fireEvent, render, screen } from "@testing-library/react";
import { Nav } from "./nav.component";
import React from "react";

describe("Nav", () => {
  test("items", () => {
    const { asFragment } = render(
      <Nav
        items={[
          {
            title: "Главная",
            link: "#",
          },
        ]}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("mobile open on humburger click", () => {
    const { asFragment } = render(
      <Nav
        items={[
          {
            title: "Главная",
            link: "#",
          },
        ]}
      />,
    );
    const firstRender = asFragment();

    fireEvent.click(screen.getByRole("button"));

    expect(firstRender).toMatchDiffSnapshot(asFragment());
  });
});
