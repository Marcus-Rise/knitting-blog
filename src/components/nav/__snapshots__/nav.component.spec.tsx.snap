// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav items 1`] = `
<DocumentFragment>
  <nav
    class="root"
  >
    <div
      class="mobileToggle"
    >
      <button
        class="root"
      >
        <img
          alt="menu"
          src="[object Object]"
        />
      </button>
    </div>
    <div
      class="itemsContainer "
    >
      <div
        class="container"
      >
        <ul
          class="ul"
        >
          <li
            class="li"
          >
            <a
              href="/#"
            >
              Главная
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Nav mobile open on humburger click 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,11 +13,14 @@
            src=\\"[object Object]\\"
          />
        </button>
      </div>
      <div
-       class=\\"itemsContainer \\"
+       class=\\"root\\"
+     />
+     <div
+       class=\\"itemsContainer d-flex\\"
      >
        <div
          class=\\"container\\"
        >
          <ul"
`;
